### DISCUSSION TITLE
Type safe

### DISCUSSION THREAD

**Discussion by robisim74** [robisim74/qwik-speak#69] [Ideas]
https://www.typescriptlang.org/play?#code/C4TwDgpgBACghsAFgZwCoHsDKwBOBLAOwHMYd0xkAeVAGliggA9gICATZKZXQogbQC6UALxRBAPhFRUDZqw5cexAFBQoAfliqoALmmyW7TugBGAKwgBjYNs0BvbWrV8A0lEJQAosxxxrlAGsIEHQAM2k6bnxicQE9VFchJkMFUwtrDVgEFAxsaJIyCmpEuj4AOgqYOhcBST1yyuqBAG5HKABfPm9cP2BA4LCIxXzY7XqW5WVQSCgAKXRCagN5TijeQToAEWWjYd5JYW0ZZJWxIU0CCAA3CBxdI52FPkJQ26gAMXOP+7VjuV3ngRXnd3nQKmUXm8AErnbTvR6rJRETIAAwAJHZ3u0MZtsXZ5otur5-FDIkiJFtxO0UbooJcbncdNo1sRWlNwNAALIgTYQCBgADy5iswCkDjUoXQ6D04qcJjgOBlbTU8oAXnoWURWk4OtrddoAI4AV0YegIRoAtiZbnrEBAADb26VQWVqMgmPAapF69qtX2TabQXn8lzBKQEgiUeBINBYJGkchUbnBwXC6ziOgAIjKmfErSAA

_Originally posted by @Zankel-Engineering in https://github.com/robisim74/qwik-speak/issues/55#issuecomment-1561702819_

**Comment by robisim74** [comment]
Hi Sebastian,

Your solution works, but there are some issues:
- The approach of this library is different: first the components are written, and from the components the `extraction `tool creates json files (whereas type safe assumes the opposite: first create the files and then write the components)
- In this library, the keys can also contain the default value (which is used by the extract tool to create the files)

Ok, the use of the extraction tool is not mandatory, and we could think about offering type safe. But how?

I could add another overload to the return function of `useTranslate`:
```
export type TranslateFn = {
  (key: string, params?: Record<string, any>, lang?: string): string;
  <V>(key: string, params?: Record<string, any>, lang?: string): V;
  <V, K>(key: K, params?: Record<string, any>, lang?: string): V;
};
```
where V is the return type, and K the key type.

But then you should always create a type for each scope in your app and pass it to each translation:

`export type homeType = Join<PathsToStringProps<typeof import('../i18n/en-US/home.json')>, ".">;`

and in component:

`t<string, homeType>('home.text')`

**EDIT**
Or better:
```
export type TranslateFn<K = string> = {
  (key: K, params?: Record<string, any>, lang?: string): string;
  <V>(key: K, params?: Record<string, any>, lang?: string): V;
};
```
and in component:
```
const t = useTranslate<homeType>();

t('home.text')
```
![image](https://github.com/robisim74/qwik-speak/assets/14012361/67aebb18-ff17-46be-8eab-307548cbb82b)

A curiosity: why are there values of type `number` in your example json? This library doesn't accept them at the moment! Of course that would be possible, but what is the use case of non-string values in translation files?

**Comment by Zankel-Engineering** [comment]
Hi Robi,

yes I understand the approach of the library. I have to figure my mind
around it, how to include the type safety for the values but keep the
original functionality.

The number in the object is just an example. Nowadays I use Chat-GPT a lot
to help me with types. It comes up with good ideas when you prompt it
correctly. It is not that far, means you need some cycles and the prompt
basis must be correct.

Another option that came to my mind was to start a file watcher to generate
an type in the background when a json file in the i18n changes and then
apply the provided utitilty types to it.

I don't know what you think about this approach? I guess it would keep the
original functionality and the computing costs should be quiet low.

So I wish you a nice thursday. Hear you soon.

Sebastian

Am Mi., 24. Mai 2023 um 23:32 Uhr schrieb Roberto Simonetti <
***@***.***>:

**Comment by robisim74** [comment]
I think we can propose the use of type safe for a scoped component in a way like this:

```typescript
import type { SpeakType } from 'qwi-speak'

export type HomeType = SpeakType<typeof import('../i18n/en-US/home.json'), '.'>;
```

```tsx
const t = useTranslate<HomeType>();

t('home.text')
```

but you need to create/update the json file _before_ writing the component (a watcher for json files doesn't seem necessary for the compiler and intellisense - at least in vscode)

_The downside is that the default value and the extraction tool could no longer be used in this case._

We should find a way to change the type to accept the default value:

`t('home.text@@My text')`

so we could modify the extraction tool to populate the keys already existing in the file with the default values if they are empty or null.

**Comment by Zankel-Engineering** [comment]
Hi Robi,

I will check that later it should be easily possible to add the default.

So you say it is checked while building the app that the keys in the
language keys match?

I guess an unittest would help a lot of people anyway because you could add
it to a git hook.

Sebastian

Roberto Simonetti ***@***.***> schrieb am Do., 25. Mai 2023,
10:25:

**Comment by robisim74** [comment]
Why a unit test? to check _if the structure of the jsons is equal over all languages_? The `Inline `plugin already generates a log, with the missing keys for each language

**Comment by Zankel-Engineering** [comment]
Hi Robi. The inline plugin would take a while to add to a githook and is
that feasible? I didn't try it out.

I have modified the type for you. You can find it here:  TypeScript: TS
Playground - An online editor for exploring TypeScript and JavaScript
(typescriptlang.org)
<https://www.typescriptlang.org/play?#code/C4TwDgpgBACghsAFgZwCoHsDKwBOBLAOwHMYd0xkAeVAGliggA9gICATZKZXQogbQC6UALxRBAPhFRUDZqw5cexAFBQoAfliqoALmmyW7TugBGAKwgBjYNs0BvbWrV8A0lEJQAosxxxrlAGsIEHQAM2k6bnxicQE9VFchJkMFUwtrDVgEFAxsaJIyCmpEuj4AOgqYOhcBST1yyuqBAG5HKABfPm9cP2BA4LCIxXzY7XqW5WVQSCgAKXRCagN5TijeQToAEWWjYd5JYW0ZZJWxIU0CCAA3CBxdI52FPkJQ26gAMXOP+7VjuV3ngRXnd3nQKmUXm8AErnbTvR6rJRETIAAwAJHZ3u0MZtsXZ5otur5-FDIkiJFtxO0UbooJcbncdNo1sRWpNptAALIgTYQCBgADy5iswCkDjUoXQ6D04qcJjgOBlbTU8oAXnoWURWk4OtrddoAI4AV0YegIRoAtiZbnrEBAADb26VQWVqMgmPAapF69qtX3s8DQXn8lzBKQEgiUeBINBYJGkchUbnBwXC6ziOgAIjKmfEbI5UAAgmBIOw8rwADJ4Fi+e1LE67TV0TAIvYxKTouyoPGYan5wNQFOhkAAdWriEwRtCoTwjHr-wUTagk+ns9bmqkAHJNwd9A2FJ3IXcAKrtMoYo9QKEQbjU2xQTvF0tscvEKs1uB14-NqczxhU887CHYIxyQFc-0oa9uB-Vd-zvNQ9CfeRXyId9bk-agYL-PMAxmQs2BfX9ZwwYCQHnFJEXyLC133Sj9ikP4KKgIIQnCOACBAe9O27DFwNneDaVI0CJyIudaGXUScKmAdOXQNg8BnCA2FIqQVIAHyLAi+MYEi+TAYdKFIugUQAARMjFNWpKTLHQAhuCgOA9Fk+TFOUvThy3SV0DKeUcB8uBVU3VobLs0UTCcuSFLwJSVNETcvP8vy1TMgAJB0nShUw8CCoA>

Best

Sebastian

Roberto Simonetti ***@***.***> schrieb am Do., 25. Mai 2023,
12:54:

**Comment by robisim74** [comment]
The Inline plugin runs during the build.  I could add an option, which if enabled breaks the build when there are missing keys.  So in your git hook, or any other CI method, just run the app build command

**Comment by Zankel-Engineering** [comment]
Yes that would be feasible the only thing that I fear is that this will take to long to execute. I can add a unittest myself no problem. I was just thinking that others could also benefit from the test when it is included in the library.

**Comment by robisim74** [comment]
Hi Sebastian,

I refactored a bit your example, to create an exportable type from the library.

It works, but not in all cases, for example when the expected return type is an [object or an array](https://robisim74.gitbook.io/qwik-speak/library/translate#arrays-and-objects-as-values)

Here the [repro](https://www.typescriptlang.org/play?#code/C4TwDgpgBAypCGBrAKuCAFewAWBnAPMgDRTpQQAewEAdgCa5S7ABOAljQOYDaAulAF4ofAHyCoyclVoMmrDpwBQUKAH5SyqAC4JU6vUYB7AEYArCAGNgm9QG9NK7gGkoHKAFEqLeFfyIIIIYAZhIkzOxcIrw6yM78lPqyJuZWarAIKGiYOASxTrwk3AB0Jegk+WI6xaXlvADcmgC+3J6sPsB+AcGhchGcUZpV9YqKoJDpEEiokABShhyEejKM4Qp8JAAiSwa9CmICmpIJy8L86jQQAG4QLNqH27LcHEE3UABiZ+93KkfSO080F63N4kEpFZ6vABKZ00bweK3kXDSAAMACS2N6NdEbLG2OCTTKzeY0fCtby+SFhRE8XgiTYiRrI7RQC7XW5aTSrLgNUZoCZTNBOAIAdTYOBgAFcgkE2BRFscdlzOCQYPDdpFxL9Eow0bYIbcAKqNIro-VQSEQZiMmxQXVG9EwY0OjLTCBCkCi8VSmVyi3MFUM5GDW3o5C4x3InljaD4gWQSXS2Xyv6yJUqtVK-a6BWyfyBELwGggG26sMO60qHSxwlukVi7AJn2EANRvnV13ZbDNqBbHMIvpZ9toOYLIeQTu5Ok9kQ8gD0ACp58p51AALJIaBtGi4AA2WDYhm3UGjTHgL2XUAAAtQALZgPfUXTIbzbh8HmjHtAX68QO8P6BbDAEBgPA3jAIYtzdBczAQHQUB5rgRQ9hAQTwBKO7AK4jAAORFNh363veWAxlA7oALSXPAO4StAuDAaBWAQUhGwoWhGFYVA2GXpe+FQPOs68uMY4QN2vYpv2CjiLh2Hpn26qcFJ3HYYOLpZFgXbENOUAAD78jWjZJsJnaiVOuqOuiSqMjOIwWIezBQDehgWIgMy4Ie4j2Co1DMDo2HIJawAyZo0HUHQOieSox4Bb5-nMEFkVQKB3ggFUfkBQAjFAti2FAIHeDeUCNI0MmcbFwAAExZTleXwAVRXYQUTREJoSXwCAADyZiWMAuBVA4WUshKN6+Zl2W5QxdXFYVzUJTlNBDb5ADMVXjflhVTY0mi8IojRzrOsDYIY6FwcY0CUTubB0IotnbphlDpVWqmQPg0bdI5zmuYeWYAETecA30NDd9mUOVj0Eq6L1oG9TkuW5NA-SFsFFH9APXXZd0UItYNxiJr0hO9sNfeI32I3QRStSAqNAxjAAs2M1pDkDQx9cMIwFSMU0U6VU+jUgAKz0xDeMOTDn3w8TFOdSkPWo4os77TAh3HVAp2uDQ52XWjt1SAAbILaCMxAzOE+LQi-RlPPa5QADs+vPcLBNi2zMFk395WW8DFAABx27jUP46LrPE6T5MsMlRSLajQA)

The best thing would be that the `SpeakType `accepts the return type as well: for example when using` t<string[]>('nested.array')` would be allowed because the expected return type is `string[]`

**Comment by Zankel-Engineering** [comment]
Hi Robi,

I played a little bit around with my example. Let me figure my mind around
how we get the return type tomorrow morning. Have to do my meditation now
:-).

TypeScript: TS Playground - An online editor for exploring TypeScript and
JavaScript (typescriptlang.org)
<https://www.typescriptlang.org/play?#code/C4TwDgpgBACghsAFgZwCoHsDKwBOBLAOwHMYd0xkAeVAGliggA9gICATZKZXQogbQC6UALxRBAPhFRUDZqw5cexAFBQoAfliqoALmmyW7TugBGAKwgBjYNs0BvbWrV8A0lEJQAosxxxrlAGsIEHQAM2k6bnxicQE9VFchJkMFUwtrDVgEFAxsaJIyCmpEuj4AOgqYOhcBST1yyuqBAG5HKABfPm9cP2BA4LCIxXzY7XqW5WVQSCgAKXRCagN5TijeQToAEWWjYd5JYW0ZZJWxIU0CCAA3CBxdI52FPkJQ26gAMXOP+7VjuV3ngRXnd3nQKmUXm8AErnbTvR6rJRETIAAwAJHZ3u0MZtsXZ5otur5-FDIkiJFtxO0UbooJcbncdNo1sRWlNwNB4EgMABBHC+ECkchUWj0E67Fn8ISiCRSP4pRH5WxaNTxBFQNJWGxqextZxuDxE3r9ELhUWS2LxRLqzXWPVOTRcnJYJFCooJGqScVPAT2pyZem3P1OergqpQGrBvSBnCtf2dI3+IKmoYW32qs5s6bQPkCp0ig5zBYEShO3n8uCCwoFugAIjKtfEbMILBwoT80AAsiBNhAIGAAPLmLVQBxqULodB6MdOExwHDT+1zgBeeklcac7Q3W+0AEcAK6MaP7gC2JluG8QEAANtep6O2mQTHg10jtxvLtwIGxF-6WNxX3yDc1HnAV6klMl8gmNQdxAisQCHdJgGQeo2jsOlT0A3gOhoNCMJPLDiA6bRoKgAB6MioB5Ng2A1JA3mTTg4HYKB6LwO4on3ax9xwCBOCvXjtEsfduHQE8XGCX8nFAytEK1FCxHtdCCEwvYiPaNpSJ3DT2RmXt+wkkApAJEt81yV1q0obt9MHYdrHEOsGybSZsyosBIHYPJeAAGTwVs4GvJZvUVXg6EwdULSkdE7FQPFMGpLMOSgGzDIAdT8xBMH3UJQjwRggv+BQIKgLKcryiKkSkAByKrC3lU5oshO4AFV2jKDEmqgKE+OAallWinl3PkLziF8-zAuasLstyxgqXauwUuCdKkFKmbKG67gprK2a+ozQaPLYEaiDG24AuoLaZuc3Scxo1a8owRaQAKhU1KIC7yuC166vVZNBmYkB+oxWKMTuxhdtpR7lsy6a8vOkqYdmxKZk7dA2DwXLv0ekRtAAH2SvswEM3GqNuhGHoJwzKEeugUQAAVpjFJWpcRidzSt8ysnsCbk+y2QokrEHQfdr1o88oCuAK8DYZRLHQAhuFkABGPQUbRjG2Cx0QqondAyjnHA9bgZcqtaWX5eAWQACYVdR9G8ExingmqnXDYNld6YACRvO8oVMPATZluWFaYABmG21ftjXHaMrWD0Yen-2AAOzeDxgABZw7th2DKdrXPxYNgyhkkAygABmToOLaYABWTP1c1qAqvz78ykTivzdkAA2OvI4bqri555D29TgB2Hvs8J3PG4HuzkLLsoVJPYeq8YAAOceo5zmPp-gwfkDKRWF9PZfZAATg3vvhNE8TgiL3fZ-30uj6XvnKMwQXhdF6BCAl68pcDjuTBFalwvtHaqV9gBiUMnfAUe8TbkUogASQIL-f+KcV6K2VlAVWWdN6T23rWROitazNAQVAZBqDpboKVtbbBtt65gNELWZuhdE6WxIWQihktpZAA>

Good Saturday for you.

Sebastian

Am Fr., 26. Mai 2023 um 18:15 Uhr schrieb Roberto Simonetti <
***@***.***>:

**Comment by Zankel-Engineering** [comment]
Ahh that was easy. I added that to the example. The other one took me two
hours to figure it out. Nothing you can really do in a company. But for a
library why not.

I added the type to the example.

Best
Sebastian

Am Sa., 27. Mai 2023 um 07:14 Uhr schrieb Sebastian Zankel <
***@***.***>:

**Comment by robisim74** [comment]
Easy?  I had tried to use ChatGPT to create this type, but I lost my patience ;)

**Comment by Zankel-Engineering** [comment]
Yes sometimes it is a bastard :-). I get angry about it from
time to time, too. I mean, it can't really reason about stuff. In the end
it is just a stupid big model of data. You have to review the output
everytime understand the type yourself and then tell it to forget
everything from before and prompt it with your last working version and
your new idea. I think it is worth putting the time into learning how to
interact with it, it really can help with some stuff when you correctly
prompt it. What I also don't like about it is that the knowledge is old.
Planned to try out Bard. Just don't find the time for it.

Am Sa., 27. Mai 2023 um 13:22 Uhr schrieb Roberto Simonetti <
***@***.***>:

**Comment by Zankel-Engineering** [comment]
Should we get that type safety into the lib? If you want I can make a draft in the next couple of days and open a pull request for you to review.

**Comment by robisim74** [comment]
Hi Sebastian,

I already created a branch: https://github.com/robisim74/qwik-speak/tree/type-safe

Here the type: https://github.com/robisim74/qwik-speak/blob/type-safe/packages/qwik-speak/src/types-safety.ts
Here the internal use: https://github.com/robisim74/qwik-speak/blob/type-safe/packages/qwik-speak/src/use-translate.ts
And here the unit test: https://github.com/robisim74/qwik-speak/blob/type-safe/packages/qwik-speak/src/tests/type-safety.test.tsx

I had to make some changes compared to what you had already done, to make arrays work even when tuple types are not available (like when importing type directly from json).

I also added return type matching.

It works, but I'm still not very happy with it: the type is very complex, and might cause intellisense to crash on long files.

**Comment by Zankel-Engineering** [comment]
Hi Robi,

what I was thinking about in the morning is if we could build a qwik speak slot that we would pass the props from the speak component down to and then get somehow the type into the scope of the child components without explicitly importing the json what I really don't like about the solution. We could add a vite plugin that uses dynamic imports to bundle a file in development which gets ditched in the prod build.

https://stackoverflow.com/questions/52112948/whats-the-return-type-of-a-dynamic-import

It seems to be impossible to infer the type into the child component functions. A dynamic import with a variable is not a suitable solution, since we know the value only to runtime. But I always say nothing is impossible. Maybe it is worth to think a little bit more about this. Maybe also you have a good idea when you figure your mind arround it how we could achieve this.

Hear you soon,
Sebastian

**Comment by Zankel-Engineering** [comment]
Hi Robi,

did you think about it? Do you think it is feasible to infer the type down?

Sebastian

**Comment by robisim74** [comment]
I haven't had the chance to think about it, the solution seems complex and this feature is not strategic: there are still changes and bugs in Qwik and so this library remains in beta version.

This is a feature for future stable releases.
